NeuralNet.js

```diff
@@ -0,0 +1,171 @@
+
+        this.model.compile({
+            optimizer: tf.train.adam(),
+            // This loss breaks things?
+            // loss: 'sparseCategoricalCrossentropy',
+            loss: 'categoricalCrossentropy',
+            metrics: ['accuracy'],
+        });
+
+        if (this.config.debug) {
+            tfvis.show.modelSummary({
+                name: 'Model Summary'
+            }, this.model);
+        }
+    }
+
+    async loadData(DATA_URL, dataType) {
+        const outputLabel = this.config.outputLabels[0];
+        const inputLabels = this.config.inputLabels;
+        console.log(inputLabels)
+
+        switch (dataType) {
+            case 'csv':
+                let data = await tf.data.csv(DATA_URL, {
+                    columnConfigs: {
+                        [outputLabel]: {
+                            isLabel: true
+                        }
+                    }
+                });
+                data = await data.toArray();
+                
+                if (this.config.debug) {
+                    const values = inputLabels.map( label => {
+                        return data.map( item => {
+                            return {
+                                x: item.xs[label],
+                                y: item.ys[outputLabel]
+                            }
+                        })
+                    })
+
+                    tfvis.render.scatterplot({
+                        name: 'debug mode'
+                    }, {
+                        values
+                    }, {
+                        xLabel: 'X',
+                        yLabel: 'Y',
+                        height: 300
+                    });
+                }
+                return data;
+            default:
+                console.log('error!')
+                return;
+        }
+    }
+
+
+    async parseData(data) {
+        return tf.tidy(() => {
+            const outputLabel = this.config.outputLabels[0];
+            const inputLabels = this.config.inputLabels;
+            // Step 1. Shuffle the data    
+            tf.util.shuffle(data);
+
+            // Step 2. Convert data to Tensor
+            // const inputs = data.map(d => inputLabels.map(header => d.xs[header]));
+            const inputs = inputLabels.map(header => data.map(d => d.xs[header]))
+            const labels = data.map(d => d.ys[outputLabel]);
+            
+            const inputTensor = tf.tensor(inputs); 
+            const labelTensor = tf.oneHot(tf.tensor1d(labels, 'int32'), this.config.NO_VAL );
+
+            //Step 3. Normalize the data to the range 0 - 1 using min-max scaling
+            const inputMax = inputTensor.max();
+            const inputMin = inputTensor.min();  
+            const labelMax = labelTensor.max();
+            const labelMin = labelTensor.min();
+
+            const normalizedInputs = inputTensor.sub(inputMin).div(inputMax.sub(inputMin)).flatten().reshape([ data.length, this.config.inputs]);
```

Comment:
This is kind of a funky maneuver, but it is a way to correctly normalize the values and reshape the data to fit the model input shape.

Reply:
We should note that this only works for numeric data and not strings.

Reply:
I wonder if we should expose a function like `normalize()` and have the user explicitly call it. I'm still confused about what makes the most sense in terms of the `NeuralNet` class having data functionality built in or that happens elsewhere and the user just passes in `xs` and `ys` to `train()`.